### DISCUSSION TITLE
Logging time

### DISCUSSION THREAD

**Discussion by Rwkeith** [HyperDbg/HyperDbg#114] [Ideas]
Hey! This is a great project and I've been reading through the well written documentation.  However, I'm trying to log time data with each print I do using the scripting engine.  I've read this portion [here](https://docs.hyperdbg.org/tips-and-tricks/misc/customize-build#configurations)

>By default, HyperDbg sends the current time of the system with each message packet from kernel-mode to user-mode, if you set ShowSystemTimeOnDebugMessages to FALSE then it no longer sends date and time along with each message.

Is this time not exposed to the scripting engine?  If not, may be a good addition to have.

Thanks

**Comment by SinaKarvandi** [comment]
Hi,
Thanks a lot for taking the time to read the documentation.
The exact meaning in documentation is for an internal logging function called `LogInfo` (that is not exposed to the script engine).

But if it matters to you, I'll add timing functions separately to the script engine, and I'll notify you on this thread.
I also add some descriptions to clarify that constant usage in the documentation.

Thanks again,

**Comment by Rwkeith** [comment]
Thanks for the response and clarification. I think this would extend the scripting interface nicely!

**Comment by SinaKarvandi** [comment]
Hi
Sorry for the very late answer.

I was thinking about the best approach we can implement such a feature. I think the best way to support it, is through a printf style format string. However, HyperDbg's script engine doesn't support TYPEs yet.

Another approach is exporting the time via a script engine function, and the user simply uses a `%s` with the result of that function. This also brings two problems. 

1. It's not a flexible approach. The user is forced to use our exported format.
2. We need to allocate a small buffer for each core so we can store the result of the time export. (of course, this is not a clean approach).

Thus, I think it's better to let script engine developers finish their tasks and add support to the structure and types. After that, we can have a clean and flexible implementation of the printf-style timing.

But let's have a solution for you so that you can use HyperDbg with timing format for now.

Go to the source code and remove the following line :
https://github.com/HyperDbg/HyperDbg/blob/cb711d347f0ca48530bf7eff4a1caedf2a5088fa/hyperdbg/include/ScriptEngineEval.h#L1747

and replace it with this code:
`    LogInfo("%s", FinalBuffer);`


Now, compile the HyperDbg, and your result of printfs (both in eval command and in events), will be changed to :
![image](https://user-images.githubusercontent.com/13383992/154972548-805c0731-50d1-4361-9b4a-f3127236aa83.png)

That's a dirty patch for you. I'll try to add this functionality to the HyperDbg whenever we finish supporting types in the HyperDbg's script engine.
